feat(a11y): Enhance accessibility for locale selection page

akshayw1 opened the pull request:
## Summary

This pull request significantly improves the accessibility of the locale selection page to comply with WCAG guidelines and provide a better experience for users relying on assistive technologies. The changes address focus management, keyboard navigation, and screen reader support while maintaining the existing functionality and code quality standards.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Accessibility improvement
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Accessibility Improvements

### Screen Reader Support
- **Enhanced semantic labels**: Screen readers now announce "Choose your language:" when entering the language list
- **Usage instructions**: Added semantic hints with clear navigation instructions
- **Position information**: Each language item announces its position (e.g., "English, 1 of 50")
- **Selection feedback**: Screen readers announce when a language is selected

###  Keyboard Navigation
- **Proper Tab cycling**: Focus cycles correctly between language list ↔ Next button
- **Home/End key support**: Jump to first/last language options
- **Arrow key navigation**: Maintained existing up/down navigation through language list
- **Enter key selection**: Select languages using Enter key
- **Type-to-search**: Preserved existing search functionality

### Focus Management
- **Auto-focus**: Language list automatically receives focus on page load
- **Focus containment**: Proper focus cycling prevents keyboard traps
- **Visual focus indicators**: Clear focus visibility throughout navigation
- **Focus traversal groups**: Logical focus order implementation

## Code Quality Improvements

- **File structure**: Moved global variables (`_nextFocusNodeProvider`) to top of file
- **Clean imports**: Added only necessary accessibility imports
- **Consistent patterns**: Follows existing accessibility patterns in the codebase
- **Error handling**: Robust keyboard event management

## Files Changed

```
packages/ubuntu_provision/lib/src/locale/locale_page.dart
```
## WCAG 2.1 Compliance

This implementation addresses the following WCAG guidelines:

- **2.1.1 Keyboard (Level A)**: All functionality accessible via keyboard
- **2.1.2 No Keyboard Trap (Level A)**: Focus moves freely between components
- **2.4.3 Focus Order (Level A)**: Logical sequential tab order
- **2.4.6 Headings and Labels (Level AA)**: Descriptive labels and instructions
- ** 2.4.7 Focus Visible (Level AA)**: Focus indicators clearly visible
- ** 3.3.2 Labels or Instructions (Level A)**: Clear usage instructions provided

## Testing Checklist

### Manual Testing

- [x] **Tab Navigation**: 
  - Language list → Next button → Language list (cycles properly)
  - Shift+Tab reverses navigation correctly
- [x] **Enhanced Keyboard Navigation**:
  - Home key jumps to first language
  - End key jumps to last language  
  - Arrow keys navigate through list
  - Enter key selects language
  - Type-to-search functionality works
- [x] **Focus Behavior**:
  - Page loads with focus on language list
  - Focus is visible throughout navigation
  - No keyboard traps or stuck focus states
- [x] **Screen Reader Testing** (tested with Orca):
  - Page title announced on load
  - "Choose your language:" announced when entering list
  - Usage instructions read appropriately
  - Position information announced for each item
  - Selection state changes announced

### Automated Testing

- [x] `flutter analyze` passes without new warnings
- [x] Existing tests continue to pass
- [x] No breaking changes to existing functionality

## Backward Compatibility

- **Fully backward compatible**: All existing functionality preserved
- **No API changes**: Uses only additive enhancements
- ** Existing behavior maintained**: Mouse interactions, selections, and navigation unchanged

## Performance Impact

- **Minimal impact**: Only adds lightweight accessibility widgets
- **No significant memory overhead**: Proper focus node management with disposal
- **Maintains responsive UI**: No performance degradation observed

By Akshay Wagmare Under Supervision of  Zendalona

ashuntu reviewed (changes requested):
My notes from https://github.com/canonical/ubuntu-desktop-provision/pull/1102 also apply here.

- Make sure all CI actions/tests pass
- Make sure to sign the CLA
- Relocate all human-readable strings to localization files

akshayw1 commented:
@ashuntu I've also submitted the CLA form, but my CI test is still failing. Could you please guide me on this?

ashuntu commented:
If the Go checks are failing, you don't need to fix them since they aren't related to your changes.

For the CLA, you need to make sure you've signed the CLA and that your commits match the email you signed the CLA with, same with Launchpad email.

ashuntu commented on packages/ubuntu_provision/lib/src/locale/locale_page.dart:
```diff
@@ -33,40 +39,63 @@ class LocalePage extends ConsumerWidget with ProvisioningPage {
+                  itemBuilder: (context, index) => Semantics(
+                    selected: index == model.selectedIndex,
+                    value:
+                        '${model.language(index)}, ${index + 1} of ${model.languageCount}',
```
"of" isn't localized, make sure to create localization files for strings like this.

akshayw1 replied:
This has been fixed, as it was an unnecessary semantic. I will ensure that similar strings are added to the localization files for other pages as well.

akshayw1 commented:
> If the Go checks are failing, you don't need to fix them since they aren't related to your changes.
> 
> For the CLA, you need to make sure you've signed the CLA and that your commits match the email you signed the CLA with, same with Launchpad email.

<img width="913" height="178" alt="image" src="https://github.com/user-attachments/assets/ad0fd4c9-c329-4a7e-819b-80b661b484be" />
still the cla is failing, any reason for this, I had first submitted form, and then logged into launchpad, Let me know if any thing there

ashuntu commented on packages/ubuntu_provision/lib/src/locale/locale_page.dart:
```diff
@@ -1,12 +1,21 @@
+/**
+ * © Akshay Waghmare
+ * Under Supervision of Zendalona
+ * Accessibility Addition
+ */
+
```
You should not need to add a copyright attribution here, it should be covered by the [license](https://github.com/canonical/ubuntu-desktop-provision/blob/ubuntu/24.04/LICENSE) of the repository and the [Ubuntu CLA](https://ubuntu.com/legal/contributors).

akshayw1 replied:
Okay, removed the same

ashuntu commented on packages/ubuntu_provision/lib/src/locale/locale_page.dart:
```diff
@@ -33,40 +39,62 @@ class LocalePage extends ConsumerWidget with ProvisioningPage {
-              await tryGetService<TelemetryService>()
-                  ?.addMetric('Language', locale.languageCode);
+          Focus(
+            onKeyEvent: (node, event) {
+              if (event is KeyDownEvent &&
+                  event.logicalKey == LogicalKeyboardKey.tab &&
+                  !HardwareKeyboard.instance.isShiftPressed) {
+                // Tab from Next button goes back to list
+                listFocusNode.requestFocus();
+                return KeyEventResult.handled;
+              }
+              return KeyEventResult.ignored;
             },
```
This change seems to work identically if this is removed, any reason this was included?

ashuntu commented on packages/ubuntu_provision/lib/src/locale/locale_page.dart:
```diff
@@ -33,40 +39,62 @@ class LocalePage extends ConsumerWidget with ProvisioningPage {
-              title: Text(model.language(index)),
-              selected: index == model.selectedIndex,
-              onTap: () => model.selectLanguage(index),
+          child: FocusTraversalGroup(
```
Here as well, this seems to function as expected without a `FocusTraversalGroup` or the `listFocusNode`.

ashuntu commented on packages/ubuntu_provision/lib/src/locale/locale_page.dart:
```diff
@@ -33,40 +39,62 @@ class LocalePage extends ConsumerWidget with ProvisioningPage {
+                  itemCount: model.languageCount,
+                  tabFocusNode: nextFocusNode,
+                  itemBuilder: (context, index) => Semantics(
+                    selected: index == model.selectedIndex,
+                    value: model.language(index),
```
You could also probably add `button: true` to this `Semantics`

ashuntu reviewed (changes requested):
You should be able to rebase on `ubuntu/24.04` now for the new CLA action.

For your changes recently:
Have you tested this locally? You still should need the `Focus` widgets that wrap the `ListWidget` and the next button, otherwise the focus doesn't properly go between the widgets.

akshayw1 commented:
> You should be able to rebase on `ubuntu/24.04` now for the new CLA action.
> 
> For your changes recently: Have you tested this locally? You still should need the `Focus` widgets that wrap the `ListWidget` and the next button, otherwise the focus doesn't properly go between the widgets.

I have updated the code and implemented the focus widget. Tested locally — it works as expected with the "Next" button and the "Choose Your Language" panel. Please review, @ashuntu.

ashuntu commented on packages/ubuntu_provision/lib/src/locale/locale_page.dart:
```diff
@@ -25,6 +27,9 @@ class LocalePage extends ConsumerWidget with ProvisioningPage {
     final lang = LocaleLocalizations.of(context);
     final nextFocusNode = ref.watch(_nextFocusNodeProvider);
 
+    // Create focus node for the language list to enable proper tab navigation
+    final listFocusNode = FocusNode();
```
This isn't being used anywhere and could be removed.

ashuntu reviewed (changes requested):
Since some of these commits represent duplicate/redundant work, please squash them into 1 or 2 where it makes sense and then we can get this merged.

akshayw1 commented:
> Since some of these commits represent duplicate/redundant work, please squash them into 1 or 2 where it makes sense and then we can get this merged.

@ashuntu, I’ve squashed all changes into a single commit. Please proceed with the merge.

akshayw1 commented:
@ashuntu , Could you please review this.

ashuntu commented:
Looks like you'll need to [sign your commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) before we can merge. Make sure to do that, and it also looks like you should rebase onto `ubuntu/24.04` again (make sure to update your fork's `ubuntu/24.04` branch for the most recent changes).

akshayw1 commented:
@ashuntu, are there any pending issues right now? If not, I think we’re good to proceed with the merge.

akshayw1 commented:
@ashuntu, could you please help me close these? I need to get them completed today as my GSoC deadline is approaching.
